### PR TITLE
Execve cgi

### PR DESCRIPTION
**Pull request by 0xCAF3D0OD**
programme cgi:
- Prend en dure un fichier binaire: bin/php-cgi et un fichier.php.
- On duplique le processus avec fork et on stock le contenu du fichier dans une std::string.

### CODE REVIEW

File: a.out
Comment (t-h2o) [inline a.out:1]: It's an extra file

File: inc/cgi.hpp [hunk 1]
```diff
@@ -0,0 +1,15 @@
+//
+// Created by Kevin Di nocera on 3/16/23.
+//
+
+#ifndef WEBSERV_CGI_HPP
+#define WEBSERV_CGI_HPP
+#include <fcntl.h>
```
Comment (t-h2o) [inline inc/cgi.hpp:7]: A little return line ?
```suggestion
#define WEBSERV_CGI_HPP

#include <fcntl.h>
```

File: inc/cgi.hpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
+//
+// Created by Kevin Di nocera on 3/16/23.
+//
+
```
Comment (t-h2o) [inline inc/cgi.hpp:4]: Nobody adds this, do we remove that ?

File: src/cgi/execve.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	if (pid == -1)
+	{
+		std::cerr << "error fork" << std::endl;
+		return (1);
+	}
+	else if (pid == 0)
+	{
+		// Child process
+		// Path to the cgi
+		arguments[0] = (char *)"/Users/kdi-noce/goinfre/bin/php-cgi";
```
[the remark this answers is not in the document]
Comment (RobinBurri) [reply src/cgi/execve.cpp:44]: from the subject pdf:
∗ The CGI should be run in the correct directory for relative path file access.
I don't know what it really means.

File: inc/cgi.hpp (hunk 2 above)
Comment (0xCAF3D0OD) [reply inc/cgi.hpp]: done

File: test.php
[the remark this answers is not in the document]
Comment (0xCAF3D0OD) [reply test.php:1]: done

[the remark this answers is not in the document]
Comment (0xCAF3D0OD) [reply test.php:1]: The comments are already in English. Those that are not in English are deleted

File: test.cpp
[the remark this answers is not in the document]
Comment (0xCAF3D0OD) [reply test.cpp:1]: deleted

File: src/cgi/execve.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+		arguments[2] = NULL;
+
+		close(pipefd[0]);
+		// Replace the old FD
+		dup2(pipefd[1], STDOUT_FILENO);
+		close(pipefd[1]);
+
+		// Execute new process
+		execve(arguments[0], arguments, envp);
+		perror("execve");
```
[the remark this answers is not in the document]
Comment (0xCAF3D0OD) [reply src/cgi/execve.cpp:57]: done

File: src/cgi/execve.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
+#include "../../inc/cgi.hpp"
```
[the remark this answers is not in the document]
Comment (0xCAF3D0OD) [reply src/cgi/execve.cpp]: done

File: a.out
Comment (0xCAF3D0OD) [reply a.out:1]: deleted

File: inc/cgi.hpp (hunk 1 above)
Comment (0xCAF3D0OD) [reply inc/cgi.hpp]: done

**Comment by t-h2o** [comment]
Do you want `cherry-pick` this commit 513ec3c0be28ae6393c0d165c42a27f2886d17cb ?

File: src/cgi/execve.cpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+			ssize_t bytes_read = read(pipefd[0], read_buffer, BUFFER_SIZE);
+			// Condition if read fail
+			if (bytes_read == -1)
+			{
+				/* EINTR error code if a signal occurred while the system call was in progress. No error
+				 * actually occurred, it's just reported that way because the system isn't able to resume the
+				 * system call automatically. This coding pattern simply retries the system call when this
+				 * happens, to ignore the interrupt.
+				 * */
+				if (errno == EINTR)
```
Comment (t-h2o) [inline src/cgi/execve.cpp:78]: [Sujet 42](https://cdn.intra.42.fr/pdf/pdf/71781/en.subject.pdf):
> **Checking the value of errno is strictly forbidden after a read or a write operation**

File: src/main.cpp [hunk 7]
```diff
@@ -1,11 +1,15 @@
 #include "arguments.hpp"
 #include "webserver.hpp"
 
+#include <fcntl.h>
+#include <stdio.h>
+#include <unistd.h>
```
Comment (t-h2o) [inline src/main.cpp:6]: It seems useless...

File: src/cgi/execve.cpp (hunk 6 above)
Comment (0xCAF3D0OD) [reply src/cgi/execve.cpp]: done 2739394

File: src/main.cpp (hunk 7 above)
Comment (0xCAF3D0OD) [reply src/main.cpp]: done af78ee0

File: src/cgi/execve.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (0xCAF3D0OD) [reply src/cgi/execve.cpp:45]: exactly

File: src/cgi/execve.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	if (pid == -1)
+	{
+		std::cerr << "error fork" << std::endl;
+		return (1);
+	}
+	else if (pid == 0)
+	{
+		// Child process
+		// Path to the cgi
+		arguments[0] = (char *)"/Users/kdi-noce/goinfre/bin/php-cgi";
```
Comment (t-h2o) [inline src/cgi/execve.cpp:45]: Where can I found `php-cgi` binary on the internet ?

Comment (0xCAF3D0OD) [reply src/cgi/execve.cpp:46]: https://learn.microsoft.com/en-us/iis/web-hosting/web-server-for-shared-hosting/fastcgi-with-php

Comment (t-h2o) [reply src/cgi/execve.cpp]: > https://learn.microsoft.com/en-us/iis/web-hosting/web-server-for-shared-hosting/fastcgi-with-php

What is this link? How can I install a `php-cgi` on my laptop?

Comment (0xCAF3D0OD) [reply src/cgi/execve.cpp]: https://www.linode.com/docs/guides/run-php-cgi-apache-ubuntu-12-04/

Comment (t-h2o) [reply src/cgi/execve.cpp]: > https://www.linode.com/docs/guides/run-php-cgi-apache-ubuntu-12-04/
> 

Do we use `apt-get` on macOS?

**Pull request closed without merging**